File: app/src/mock/java/br/com/mobicare/cielo/meuCadastro/data/clients/managers/MeuCadastroRepository.java
package br.com.mobicare.cielo.meuCadastro.data.clients.managers;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;

import androidx.annotation.NonNull;

import br.com.mobicare.cielo.commons.utils.ReaderMock;
import br.com.mobicare.cielo.meuCadastro.domains.entities.CardBrandFees;
import br.com.mobicare.cielo.meuCadastro.domains.entities.MeuCadastroEndereco;
import br.com.mobicare.cielo.meuCadastro.domains.entities.MeuCadastroObj;
import br.com.mobicare.cielo.meuCadastro.presetantion.ui.MeuCadastroCallback;

/**
 * Created by silvia.miranda on 25/04/2017.
 */

public class MeuCadastroRepository {
    private Context context;

    public static MeuCadastroRepository getInstance(@NonNull Context context) {
        return new MeuCadastroRepository(context);
    }

    public MeuCadastroRepository(@NonNull Context context) {
        this.context = context;
    }

    public void getMeuCadastro(final MeuCadastroCallback callback){
        callback.onStart();

        //Espera 3 segundos para enviar a resposta de sucesso
        new Handler(Looper.getMainLooper()).postDelayed(new Runnable() { // Tried new Handler(Looper.myLopper()) also
            @Override
            public void run() {

                MeuCadastroObj obj = ReaderMock.Companion.getMeuCadastro(context);
//                MeuCadastroObj obj = ReaderMock.Companion.getMeuCadastroUnico(context);
                callback.onSuccess(obj);

                MeuCadastroEndereco endFisico = obj.getEndereco(MeuCadastroEndereco.Tipo.FISICO);
                MeuCadastroEndereco endContato = obj.getEndereco(MeuCadastroEndereco.Tipo.CONTATO);

                if (endFisico == null && endContato == null) {
                    //Caso nao tenha nenhum dos dois enderecos retorna error
                    callback.onLoadContactAddress(null);
                    callback.onLoadPhysicalAddress(null);
                } else if (endFisico == null) {
                    //Caso nao tenha o endereco fisico
                    callback.onLoadContactAddress("OK");
                    callback.onLoadPhysicalAddress(null);
                } else if (endContato == null) {
                    //Caso nao tenha o endereco de contato
                    callback.onLoadPhysicalAddress("OK");
                    callback.onLoadContactAddress(null);
                } else {
                    //Caso tenham os dois enderecos
                    callback.onLoadPhysicalAddress("OK");
                    callback.onLoadContactAddress("Error");
                }
//                callback.onError(new ErrorMessage());
                callback.onFinish();
            }
        }, 3 * 1000);
    }

    public void getBrands(final MeuCadastroCallback callback){
        callback.onStart();

        //Espera 3 segundos para enviar a resposta de sucesso
        new Handler(Looper.getMainLooper()).postDelayed(new Runnable() { // Tried new Handler(Looper.myLopper()) also
            @Override
            public void run() {

                CardBrandFees obj = ReaderMock.Companion.getBrands(context);
//                MeuCadastroObj obj = ReaderMock.Companion.getMeuCadastroUnico(context);
                callback.onSuccessBrands(obj);
                //callback.onErrorBrands(new ErrorMessage());

//                callback.onError(new ErrorMessage());
                callback.onFinish();
            }
        }, 3 * 1000);
    }
}

File: app/src/api/java/br/com/mobicare/cielo/meuCadastro/data/clients/managers/MeuCadastroRepository.kt
package br.com.mobicare.cielo.meuCadastro.data.clients.managers

import android.annotation.SuppressLint
import br.com.mobicare.cielo.meuCadastro.data.clients.api.APICallbackPassword
import br.com.mobicare.cielo.meuCadastro.data.clients.api.MeuCadastroAPIDataSource
import br.com.mobicare.cielo.meuCadastro.domains.entities.MeuCadastroEndereco
import br.com.mobicare.cielo.meuCadastro.domains.entities.MeuCadastroObj
import br.com.mobicare.cielo.meuCadastro.presetantion.ui.MeuCadastroCallback
import br.com.mobicare.cielo.meusCartoes.clients.api.domain.BodyChangePassword
import com.google.gson.internal.`$Gson$Preconditions`.checkNotNull
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers

/**
 * Created by silvia.miranda on 25/04/2017.
 */

open class MeuCadastroRepository(remoteDataSource: MeuCadastroAPIDataSource) {
    private val remoteDataSource: MeuCadastroAPIDataSource = checkNotNull(remoteDataSource)
    private var ERROR = "ERROR"

    //TODO: Refazer
    @SuppressLint("CheckResult")
    fun getChangePassword(body: BodyChangePassword, callback: APICallbackPassword) {
        remoteDataSource.getChangePassword(body)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ data ->
                    when {
                        data.code() in 200..204 -> {
                            callback.onSuccess()
                        }
                        data.code() in 400..420 -> {
                            callback.onErrorAuthentication(data.message())
                        }
                        else -> {
                            callback.onError(data.message())
                        }
                    }
                },
                        { error ->
                            val erro = error?.message ?: ""
                            callback.onError(erro)
                        })
    }

    private fun checkAddress(response: MeuCadastroObj, callback: MeuCadastroCallback) {
        if (response.addresses != null) {
            val endFisico = response.getEndereco(MeuCadastroEndereco.Tipo.FISICO)
            val endContato = response.getEndereco(MeuCadastroEndereco.Tipo.CONTATO)

            if (endFisico == null && endContato == null) {
                //Caso nao tenha nenhum dos dois enderecos retorna error
                callback.onLoadContactAddress(null)
                callback.onLoadPhysicalAddress(null)
            } else if (endFisico == null) {
                //Caso nao tenha o endereco fisico
                getMapaUrl(endContato?.addressConcatenadoMap, MeuCadastroEndereco.Tipo.CONTATO, callback)
                callback.onLoadPhysicalAddress(null)
            } else if (endContato == null) {
                //Caso nao tenha o endereco de contato
                getMapaUrl(endFisico?.addressConcatenadoMap, MeuCadastroEndereco.Tipo.FISICO, callback)
                callback.onLoadContactAddress(null)
            } else {
                //Caso tenham os dois enderecos
                getMapaUrl(endFisico?.addressConcatenadoMap, MeuCadastroEndereco.Tipo.FISICO, callback)
                getMapaUrl(endContato?.addressConcatenadoMap, MeuCadastroEndereco.Tipo.CONTATO, callback)
            }
        }
    }

    fun getMapaUrl(address: String?, type: String, callback: MeuCadastroCallback) {
        remoteDataSource.getMap(address)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(object : Observer<MeuCadastroEndereco> {
                    override fun onSubscribe(d: Disposable) {

                    }

                    override fun onComplete() {

                    }

                    override fun onError(e: Throwable) {
                        when (type) {
                            MeuCadastroEndereco.Tipo.FISICO -> callback.onLoadPhysicalAddress(ERROR)
                            MeuCadastroEndereco.Tipo.CONTATO -> callback.onLoadContactAddress(ERROR)
                        }
                    }


                    override fun onNext(response: MeuCadastroEndereco) {
                        when (type) {
                            MeuCadastroEndereco.Tipo.FISICO -> callback.onLoadPhysicalAddress(response.status)
                            MeuCadastroEndereco.Tipo.CONTATO -> callback.onLoadContactAddress(response.status)
                        }
                    }
                })
    }


    companion object {
        fun getInstance(remoteDataSource: MeuCadastroAPIDataSource): MeuCadastroRepository = MeuCadastroRepository(remoteDataSource)
    }
}
